### PR TITLE
Added analytics tab to webspace settings

### PR DESCRIPTION
**Pull request by wachterjohannes**
__tasks:__

- [x] test coverage

__informations:__

| q                | a
| ---------------- | ---
| Fixed tickets    | none
| Related PRs      | https://github.com/sulu-io/sulu-standard/pull/585  https://github.com/massiveart/MassiveSearchBundle/pull/73 https://github.com/massiveart/husky/pull/595
| BC breaks        | none
| Documentation PR | https://github.com/sulu-io/sulu-docs/pull/146

### CODE REVIEW

**Comment by wachterjohannes** [comment]
Example piwik tracking code generated by a local test-installation:

```js
<!-- Piwik -->
<script type="text/javascript">
  var _paq = _paq || [];
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u="//piwik.lo/";
    _paq.push(['setTrackerUrl', u+'piwik.php']);
    _paq.push(['setSiteId', 1]);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'piwik.js'; s.parentNode.insertBefore(g,s);
  })();
</script>
<noscript><p><img src="//piwik.lo/piwik.php?idsite=1" style="border:0;" alt="" /></p></noscript>
<!-- End Piwik Code -->
```

**Comment by wachterjohannes** [comment]
@danrot this is ready to review. style-ci fails in all branches and should be fixed in develop directly. therefor i have made a issue https://github.com/sulu-io/sulu/issues/1914 and a pr.

**Comment by wachterjohannes** [comment]
@danrot i have found some issues with @berndhep i will fix it an mention you when i am finished

**Comment by danrot** [comment]
:+1:

**Comment by wachterjohannes** [comment]
@danrot should be good now (: i only have to fix the style of the checkbox

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $analytic = $event->getObject();
+
+        if (!($analytic instanceof Analytic)) {
+            return;
+        }
+
+        if ($analytic->getAllDomains()) {
+            $metadata = new PropertyMetadata($event->getType()['name'], 'domains');
+            $value = new \stdClass();
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php]: i have not found a better solution to override the `domains` property with other data.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,10 +2,13 @@ CHANGELOG for Sulu
 ==================
 
 * dev-develop
+    * FEATURE     #1912 [WebsiteBundle]   Added analytics to webspace settings
     * BUGFIX      #1874 [ContentBundle]   Fixed preview selector for blocks
     * FEATURE     #1777 [ContentBundle]   Added selection content types
     * BUGFIX      #1911 [SecurityBundle]  Fixed default locale user builder
     * BUGFIX      #1915 [All]             Removed deprecations of initial admin request
+    * FEATURE     #1851 [SnippetBundle]   Added default snippets webspace-settings
+    * FEATURE     #1851 [ContentBundle]   Added webspace-settings
```
Comment (danrot) [inline CHANGELOG.md]: Why are these lines added?

File: src/Sulu/Bundle/AdminBundle/Resources/translations/sulu/backend.en.xlf [hunk 3]
```diff
@@ -920,6 +920,14 @@
...
                 <target>Show all</target>
             </trans-unit>
+            <trans-unit id="c229c52dae1354cc4123c0b4820e23cdb" resname="public.type">
+                <source>public.type</source>
+                <target>Type</target>
+            </trans-unit>
+            <trans-unit id="c129c52dae1354cc4123c0b4820e23cdb" resname="labels.success.save-desc">
+                <source>labels.success.save-desc</source>
+                <target>Sucessfully saved</target>
+            </trans-unit>
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/translations/sulu/backend.en.xlf]: Didn't you also ask for french translations?

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php [hunk 4]
```diff
@@ -69,5 +69,13 @@ public function onPostSerialize(ObjectEvent $event)
...
         $portalInformation = $event->getContext()->accept(array_values($portalInformation));
         $visitor->addData('portalInformation', $portalInformation);
+
+        $urls = [];
+        foreach ($webspace->getPortals() as $portal) {
+            $environment = $portal->getEnvironment($this->environment);
+            $urls = array_merge($urls, $environment->getUrls());
+        }
+        $urls = $event->getContext()->accept($urls);
+        $visitor->addData('urls', $urls);
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: Why don't you add a `getUrls` method to the webspace, and use a virtual property? Then this functionality would also be available somewhere else, in case we need it.

File: src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    private $repository;
+
+    /**
+     * @var DomainRepository
+     */
+    private $domainRepository;
+
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        AnalyticRepository $repository,
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: Would call this class `AnalyticsRepository`, to keep the same term everywhere.

Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: And wouldn't you call the variable `$analyticsRepository` to avoid confusion, in case there'll be more repositories injected at some point?

File: src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function removeMultiple($ids)
+    {
+        foreach ($ids as $id) {
+            $this->entityManager->remove($this->entityManager->getReference(Analytic::class, $id));
+        }
+
+        $this->entityManager->flush();
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: Didn't we say that we should flush at the very end (meaning the controller or the command)? Then this method would probably be not necessary. /cc @chirimoya

File: src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+     */
+    private function setData(Analytic $analytic, $webspaceKey, $data)
+    {
+        $analytic->setTitle($this->getValue($data, 'title'));
+        $analytic->setType($this->getValue($data, 'type'));
+        $analytic->setContent($this->getValue($data, 'content', ''));
+        $analytic->setAllDomains($this->getValue($data, 'allDomains', false));
+        $analytic->setWebspaceKey($webspaceKey);
+
+        $analytic->clearDomains();
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: Same as before, I think this variable and the class should be called `Analytics`.

File: src/Sulu/Bundle/WebsiteBundle/Controller/AnalyticsController.php [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+use Hateoas\Representation\CollectionRepresentation;
+use Hateoas\Representation\RouteAwareRepresentation;
+use Sulu\Bundle\ContentBundle\Admin\ContentAdmin;
+use Sulu\Component\Rest\RestController;
+use Sulu\Component\Security\SecuredControllerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Provides webspace analytic rest-endpoint.
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Controller/AnalyticsController.php]: `analytics`

File: src/Sulu/Bundle/WebsiteBundle/Entity/Analytic.php [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+namespace Sulu\Bundle\WebsiteBundle\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+
+/**
+ * Analytic.
+ */
+class Analytic
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Entity/Analytic.php]: Maybe we can call this one also `Analytics`? Or are there any problems with the plural?

File: src/Sulu/Bundle/WebsiteBundle/Entity/AnalyticRepository.php [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Returns analytic-key by url.
+     *
+     * @param string $urlExpression
+     * @param string $environment
+     *
+     * @return Analytic[]
+     */
+    public function findByUrl($urlExpression, $environment)
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Entity/AnalyticRepository.php]: Why is this variable called `$urlExpression`, and not just `$url`?

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+    public function onPostSerialize(ObjectEvent $event)
+    {
+        $analytic = $event->getObject();
+
+        if (!($analytic instanceof Analytic)) {
+            return;
+        }
+
+        if ($analytic->getAllDomains()) {
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php]: This method is returning a boolean, right? Then the name is a bit confusing, I was expecting to get an array of `Domain` objects.

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php (hunk 1 above)
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php]: I'd say this is ok for serialization purposes, but why do you need that at all?

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AppendAnalyticsListener.php [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        if ($event->getRequest()->getRequestFormat() !== 'html') {
+            return;
+        }
+
+        $analytics = $this->analyticRepository->findByUrl(
+            $this->requestAnalyzer->getPortalInformation()->getUrlExpression(),
+            $this->environment
+        );
+
+        $content = $this->engine->render('SuluWebsiteBundle:Analytics:render.html.twig', ['analytics' => $analytics]);
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AppendAnalyticsListener.php]: Can't you name the template a bit better? `render` could mean anything.

File: src/Sulu/Bundle/WebsiteBundle/Resources/public/js/components/webspace/settings/analytics/main.js [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+                    event = 'husky.datagrid.records.change';
+                }
+
+                this.sandbox.emit(event, response);
+                this.sandbox.emit('sulu.labels.success.show', this.translations.successMessage, this.translations.successLabel);
+            }.bind(this));
+        },
+
+        deleteAnalytics: function() {
+            var ids = JSON.parse(JSON.stringify($('#webspace-analytics-list').data('selected')));
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resources/public/js/components/webspace/settings/analytics/main.js]: I'd say it would be more elegant to use the `this.sandbox.util.deepCopy` method instead

File: src/Sulu/Bundle/WebsiteBundle/Resources/public/js/components/webspace/settings/analytics/main.js [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+            this.sandbox.sulu.showDeleteDialog(function(confirmed) {
+                if (!!confirmed) {
+                    this.sandbox.util.save(
+                        this.templates.url({webspace: this.data, id: null, ids: ids}), 'DELETE'
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resources/public/js/components/webspace/settings/analytics/main.js]: `DELETE` goes into a new line

File: src/Sulu/Bundle/WebsiteBundle/Resources/public/js/components/webspace/settings/analytics/overlay/main.js [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+            $(contentSelector).html(
+                _.template(type.inputTemplate, {
+                        labels: _.map(type.labels, function(item) {
+                            return this.sandbox.translate(item);
+                        }.bind(this))
+                    }
+                )
+            );
+
+            var defs = [];
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resources/public/js/components/webspace/settings/analytics/overlay/main.js]: `deferreds`

File: src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php (hunk 7 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: but analytic is an entity we dont use plural for entities only for repositories i think.

File: src/Sulu/Bundle/WebsiteBundle/Entity/Analytic.php (hunk 9 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Entity/Analytic.php]: ne problems with plurals but we dont use plurals for entities (see contact or account)

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php (hunk 1 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php]: i would like to return `true` here if the flag `allDomains` isset. this can be used to display "all domains" in the datagrid.

File: CHANGELOG.md (hunk 2 above)
Comment (wachterjohannes) [inline CHANGELOG.md]: because i have forgotten it in the PR before

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php (hunk 11 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticSerializeEventSubscriber.php]: should i call it `isAllDomains`?

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: its not really possible because we need the environment here. to achieve that we would have to inject the environment into the webspace. should i do that?

File: src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: we dont use plural for Entities and repositories. see `Contact` and `ContactRepository`

Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: maybe i should also rename `AnalyticsManager` to `AnalyticManager` because of `ContactManager`

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AppendAnalyticsListener.php (hunk 12 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AppendAnalyticsListener.php]: fixed

File: src/Sulu/Bundle/AdminBundle/Resources/translations/sulu/backend.en.xlf (hunk 3 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/AdminBundle/Resources/translations/sulu/backend.en.xlf]: i have added this two missing translations

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php (hunk 4 above)
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: Ok, I think we can leave that for now. However, we should probably try to get rid of this state at some point.

File: src/Sulu/Bundle/WebsiteBundle/Entity/Analytic.php (hunk 9 above)
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Entity/Analytic.php]: My point is that all the services already have Analytics in their name, and not Analytic (e.g. Google Analytics), so the singular sounds strange to me here.

**Comment by danrot** [comment]
@wachterjohannes Tests are failing

**Comment by wachterjohannes** [comment]
@danrot the tests should be green when the PR (https://github.com/massiveart/MassiveSearchBundle/pull/73) in search-bundle is merged

File: src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+    {
+        foreach ($ids as $id) {
+            $this->entityManager->remove($this->entityManager->getReference(Analytics::class, $id));
+        }
+    }
+
+    /**
+     * Set data to given key.
+     *
+     * @param Analytics $analytic
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Analytics/AnalyticsManager.php]: The variable should also be called `$analytics`

File: src/Sulu/Bundle/WebsiteBundle/Entity/AnalyticsRepository.php [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+            ->where('a.webspaceKey = :webspaceKey');
+
+        $query = $queryBuilder->getQuery();
+        $query->setParameter('webspaceKey', $webspaceKey);
+
+        return $query->getResult();
+    }
+
+    /**
+     * Returns analytic-key by id.
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Entity/AnalyticsRepository.php]: `analytics-key`

File: src/Sulu/Bundle/WebsiteBundle/Entity/AnalyticsRepository.php [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+            ->where('a.id = :id');
+
+        $query = $queryBuilder->getQuery();
+        $query->setParameter('id', $id);
+
+        return $query->getSingleResult();
+    }
+
+    /**
+     * Returns analytic-key by url.
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Entity/AnalyticsRepository.php]: `analytics-key`

File: src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticsSerializeEventSubscriber.php [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+                'event' => Events::POST_SERIALIZE,
+                'format' => 'json',
+                'method' => 'onPostSerialize',
+            ],
+        ];
+    }
+
+    public function onPostSerialize(ObjectEvent $event)
+    {
+        $analytic = $event->getObject();
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/EventListener/AnalyticsSerializeEventSubscriber.php]: `$analytics`

**Comment by danrot** [comment]
@wachterjohannes The first time I open the analytics tab it works, but the request is sent to the URL `http://sulu.lo/admin/api/webspaces/sulu_io/analytics?fields=,,&limit=50`. The field parameter looks a bit strange.

**Comment by danrot** [comment]
If I press the add button, google is "somehow" preselected in the select, but somehow not. The select says "Please choose", but if I click on it "Google" is marked as selected, and I can't choose it without changing the selection first. If I still click ok in the overlay the subsequent request fails.

**Comment by danrot** [comment]
What kind of validation is done in the fields for piwik? I don't know what to enter :confused:

**Comment by danrot** [comment]
Do you also purge the HTTP cache after changing an entry of the analytics key? Otherwise the changes won't have the desired effect.

**Comment by wachterjohannes** [comment]
you have to fill in string greater 3 and an integer. maybe we should remove this validation it feels a bit strange ...

**Comment by danrot** [comment]
I guess that's the site id, but what should be in the other field? If I enter some random URL it doesn't work.

**Comment by wachterjohannes** [comment]
you have to fill in in the first element an string with length > 3 and in the second one a integer both have to be valid until the red labels disappears.

**Comment by wachterjohannes** [comment]
i have to change the select in husky to prevent select first element take a look at this PR https://github.com/massiveart/husky/pull/595

**Comment by wachterjohannes** [comment]
@danrot fixed all comments except for the validation. we should talk about this topic later

File: src/Sulu/Bundle/WebsiteBundle/DependencyInjection/Configuration.php [hunk 20]
```diff
@@ -36,6 +36,12 @@ public function getConfigTreeBuilder()
                     ->scalarNode('analytics_key')->defaultValue('UA-SULU-PREVIEW-KEY')->end()
                 ->end()
             ->end()
+            ->arrayNode('analytics')
+                ->addDefaultsIfNotSet()
+                ->children()
+                    ->booleanNode('enabled')->defaultTrue()->end()
+                ->end()
+            ->end()
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/DependencyInjection/Configuration.php]: Why don't you use the `canBeDisabled` call as described in http://symfony.com/doc/current/components/config/definition.html#optional-sections ?

**Comment by danrot** [comment]
And isn't it possible to at least validate the piwik fields on their own? It's very confusing if you don't know which field is wrong.

**Comment by danrot** [comment]
The overlay still says Analytic instead of Analytics:
image

**Comment by wachterjohannes** [comment]
i have created an issue for piwik validation https://github.com/sulu-io/sulu/issues/1941

File: src/Sulu/Bundle/WebsiteBundle/Resources/views/Analytics/type/piwik.html.twig [hunk 21]
```diff
@@ -0,0 +1,13 @@
+<script type="text/javascript">
+    var _paq = _paq || [];
+    _paq.push(['trackPageView']);
+    _paq.push(['enableLinkTracking']);
+    (function() {
+        var u="{{ analytics.content.url }}";
+        _paq.push(['setTrackerUrl', u+'piwik.php']);
+        _paq.push(['setSiteId', {{ analytics.content.siteId }}]);
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Resources/views/Analytics/type/piwik.html.twig]: Don't you have to put quotes around this, if the siteId can also be a UUID?
